### ISSUE TITLE
Trait sealing: Items with `where Self: Sized` bounds cannot seal

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#1346]
Note this example in https://github.com/rust-lang/rust/issues/57900:
```rust
trait Foo {
    fn bar(&self) -> Self where Self: Sized;
}

impl Foo for str {}
```

`str` is of course built-in so orphan rules would prevent this if `Foo` were from another crate. But `dyn Downstream` is also unsized, and if `Downstream` is in our crate then I believe `impl Foo for dyn Downstream` should be a possible example of an `impl Foo` that downstream can provide.

So even if the `bar` item above had an unnamable type in its signature, that wouldn't fully seal the trait.

This might cause some false-positives around traits that use `Self: Sized` bounds on items.
